### PR TITLE
chore: add typechecking on i18n layer

### PR DESCRIPTION
**Pull request by pdesoyres-cc**
## What is this PR?

This PR adds type-checking on the i18n layer.

It also adds some minor improvements:

* All related files have been moved to a new `src/lib/i18n` directory.
* Test files have been moved accordingly.
* `i18n-date.js` has been reworked to make date formatting faster (the `new Intl.DateTimeFormat(...)` is made once)

Additionally, it adds a type-checking on our translations. We now have two layers:

* one generic i18n layer (everything in `src/lib/i18n`)
* one specific layer with strong typing on our translations (everything in `src/translations`)

This means that TS will check that 
* you use a valid translation key
* you use the right parameter when the translation is a function
* you don't use any parameter when the translation is a string

This is great, but now, to take advantage of that, we have a lot of work for typing the translate functions in `translation.*.js` files.

## How to review?

* Check the code
* Check that wrong usage off the `i18n` function will be detected by TS
* Check that all tests pass
* Check stories with Storybook
* If you are lost on how to review, call me!

### CODE REVIEW

File: src/lib/i18n/i18n-sanitize.js [hunk 1]
```diff
@@ -64,7 +83,7 @@ export function sanitize(statics, ...params) {
         });
 
       // If link has href and external origin => force rel and target
-      if (node.tagName === 'A' && node.getAttribute('href') != null) {
+      if (isLinkNodeWithHref(node)) {
```
Comment (florian-sanders-cc) [inline src/lib/i18n/i18n-sanitize.js:86]: **suggestion:** I've done some research because it's not the first time I face this situation. Here's what I got:
```suggestion
      if (node instanceof HTMLAnchorElement) {
```

WDYT?

File: src/lib/i18n/i18n-sanitize.js [hunk 2]
```diff
@@ -24,15 +33,25 @@ const escapeHtml = (() => {
...
-    sibling.parentNode.removeChild(sibling);
-    return sibling.data;
+/**
+ * If the sibling node is a text node, get text content and remove if from the DOM.
+ *
+ * @param {Node|null} siblingNode
+ * @return {string}
+ */
+function absorbTextSibling(siblingNode) {
+  if (siblingNode != null && isTextNode(siblingNode)) {
```
Comment (florian-sanders-cc) [inline src/lib/i18n/i18n-sanitize.js:43]: **suggestion:** 
```suggestion
  if (siblingNode != null && siblingNode instanceof Text) {
```

With this we may even get rid of the `siblingNode != null` :thinking:

File: src/lib/i18n/i18n.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
+/**
+ * @typedef {import('./i18n.types.d.ts').TranslationsMap} TranslationsMap
+ * @typedef {import('./i18n.types.d.ts').Translations} Translations
+ * @typedef {import('./i18n.types.d.ts').TranslateFunction} TranslateFunction
+ * @typedef {import('./i18n.types.d.ts').Translated} Translated
```
Comment (florian-sanders-cc) [inline src/lib/i18n/i18n.js:5]: **nitpick:**
```suggestion
 * @typedef {import('./i18n.types.js').TranslationsMap} TranslationsMap
 * @typedef {import('./i18n.types.js').Translations} Translations
 * @typedef {import('./i18n.types.js').TranslateFunction} TranslateFunction
 * @typedef {import('./i18n.types.js').Translated} Translated
```

File: src/translations/translation.types.d.ts [hunk 4]
```diff
@@ -0,0 +1,14 @@
+import { translations as enTranslations } from './translations.en.js';
+import { translations as frTranslations } from './translations.fr.js';
+
+type Translations = Omit<typeof frTranslations & typeof enTranslations, 'LANGUAGE'>;
+type TranslationKeys = keyof Translations;
+type Function = (...args: any) => any;
```
Comment (florian-sanders-cc) [inline src/translations/translation.types.d.ts:6]: **suggestion:** I didn't think of it when we create that but since `Function` actually exists, maybe we should go another name?

```suggestion
type StandardArrowFunction = (...args: any) => any;
```

```suggestion
type ArrowFunctionWithArgs = (...args: any) => any;
```

(none of these names convince me, I'll try to think of something better...)

File: src/translations/translations.en.js [hunk 5]
```diff
@@ -1,25 +1,33 @@
...
-} from '../lib/i18n-number.js';
-import { sanitize } from '../lib/i18n-sanitize.js';
-import { preparePlural } from '../lib/i18n-string.js';
+} from '../lib/i18n/i18n-number.js';
+import { sanitize } from '../lib/i18n/i18n-sanitize.js';
+import { preparePlural } from '../lib/i18n/i18n-string.js';
+
+/**
+ * @typedef {import('../lib/i18n/i18n.types.d.ts').Translations} Translations
+ */
```
Comment (florian-sanders-cc) [inline src/translations/translations.en.js:21]: **nitpick:** we don't use it anymore

File: src/translations/translations.fr.js [hunk 6]
```diff
@@ -1,20 +1,24 @@
...
-} from '../lib/i18n-number.js';
-import { sanitize } from '../lib/i18n-sanitize.js';
-import { preparePlural } from '../lib/i18n-string.js';
+} from '../lib/i18n/i18n-number.js';
+import { sanitize } from '../lib/i18n/i18n-sanitize.js';
+import { preparePlural } from '../lib/i18n/i18n-string.js';
+
+/**
+ * @typedef {import('../lib/i18n/i18n.types.d.ts').Translations} Translations
+ */
```
Comment (florian-sanders-cc) [inline src/translations/translations.fr.js:21]: **nitpick:** we don't use it anymore
```suggestion
```

**Review by florian-sanders-cc — CHANGES_REQUESTED** [review]
Great job @pdesoyres-cc I can't wait for these changes to be merged :tada: 

Officially this is not a breaking change but we'll have to keep in mind that the console sometimes relies on i18n and that its imports should be updated there as well when using the release with these changes :+1:

**Review by florian-sanders-cc — APPROVED** [review]
Great job, thanks again! Can't wait for this to get merged :eyes:

**Review by HeleneAmouzou — APPROVED** [review]
Thanks and GG :clap: !
I only have noted few typos when I was checking the code, but otherwise I have made some tests with wrong usages of the `i18n` function and checked some stories and it looks good to me !

**Review by roberttran-cc — APPROVED** [review]
Awesome PR, great job!

**Review by Galimede — APPROVED** [review]
LGTM, thanks for this, that's nice! :muscle:

File: .storybook/preview.js [hunk 7]
```diff
@@ -21,9 +20,11 @@ Array
...
   });
 
-const availableLanguages = Object
-  .entries(getAvailableLanguages())
-  .map(([title, value]) => ({ value, title }));
+const availableLanguages = [
+  {value: 'en', title: 'English'},
+  {value: 'fr', title: 'Français'},
+  {value: 'missing', title:'🤬 Missing'},
+];
```
Comment (hsablonniere) [inline .storybook/preview.js:27]: nitpick: prettier adds spaces at the beginning and the end of object declarations

:arrow_right: this change is available in my proposition branch: https://github.com/CleverCloud/clever-components/tree/typecheck-i18n--hubert

File: docs/translations.example.js [hunk 8]
```diff
@@ -20,10 +25,10 @@ const unknownTagFunction = () => ``;
...
 // Translations must be grouped with fold region comments like this:
-//#region cc-component
-//#endregion
+/**
+ * @type {Translations}
+ */
 export const translations = {
-
   // First, declare the translated language name with some emoji flag with "LANGUAGE"
   LANGUAGE: 'Français',
```
Comment (hsablonniere) [inline docs/translations.example.js:33]: issue: I think we decided we would remove the `LANGUAGE` key as we no longer need it and it brings more problems than solutions

:arrow_right: this change is available in my proposition branch: https://github.com/CleverCloud/clever-components/tree/typecheck-i18n--hubert

File: src/components/cc-beta/cc-beta.types.d.ts [hunk 9]
```diff
@@ -1 +1 @@
-type PositionType = 'top-left' | 'bottom-left' | 'top-right' | 'bottom-right';
+export type PositionType = 'top-left' | 'bottom-left' | 'top-right' | 'bottom-right';
```
Comment (hsablonniere) [inline src/components/cc-beta/cc-beta.types.d.ts:1]: issue: this should be in another commit

:arrow_right: this change is available in my proposition branch: https://github.com/CleverCloud/clever-components/tree/typecheck-i18n--hubert

File: src/translations/translation.types.d.ts [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+type TranslationKeys = keyof Translations;
+type ArrowFunction = (...args: any) => any;
+
+export type I18nFunction = <Key extends TranslationKeys, TranslationPropertyValue extends Translations[Key]>(
+  key: Key,
+  // the rest of the arguments depends on the type of the TranslationPropertyValue which can be either a function or a string (see (i18n.types.d.ts).Translation)
+  // if it's a string, we do not allow any arguments
+  // if it's a function, we allow one argument with the same type as the one expected by the function
+  ...args: TranslationPropertyValue extends ArrowFunction ? [Parameters<TranslationPropertyValue>[0]] : [undefined]
+) => TranslationPropertyValue extends ArrowFunction ? ReturnType<TranslationPropertyValue> : TranslationPropertyValue;
```
Comment (hsablonniere) [inline src/translations/translation.types.d.ts:14]: praise: impressive work!!! :clap:

File: src/translations/translations.en.js [hunk 11]
```diff
@@ -34,16 +42,24 @@ const formatBytes = prepareNumberBytesFormatter(lang, 'B', ' ');
...
+ * @param {Flavor} flavor
+ * @return {string}
+ */
+function formatFlavor(flavor) {
+  const cpu = `CPUs: ${flavor.cpus}`;
+  const shared = flavor.microservice ? ` (shared)` : '';
+  const gpu = flavor.gpus > 0 ? `GPUs: ${flavor.gpus}` : '';
+  const mem = `RAM: ${formatBytes(flavor.mem * 1024 * 1024)}`;
   return [cpu + shared, gpu, mem].filter((a) => a).join('\n');
 }
```
Comment (hsablonniere) [inline src/translations/translations.en.js:65]: issue: I think we decided we would remove the `LANGUAGE` key as we no longer need it and it brings more problems than solutions

:arrow_right: this change is available in my proposition branch: https://github.com/CleverCloud/clever-components/tree/typecheck-i18n--hubert

File: src/translations/translations.fr.js [hunk 12]
```diff
@@ -45,16 +53,24 @@ const formatBytes = prepareNumberBytesFormatter(lang, 'o', '\u202f');
...
+ * @param {Flavor} flavor
+ * @return {string}
+ */
+function formatFlavor(flavor) {
+  const cpu = `CPUs : ${flavor.cpus}`;
+  const shared = flavor.microservice ? ` (partagé)` : '';
+  const gpu = flavor.gpus > 0 ? `GPUs : ${flavor.gpus}` : '';
+  const mem = `RAM : ${formatBytes(flavor.mem * 1024 * 1024)}`;
   return [cpu + shared, gpu, mem].filter((a) => a).join('\n');
 }
```
Comment (hsablonniere) [inline src/translations/translations.fr.js:76]: issue: I think we decided we would remove the `LANGUAGE` key as we no longer need it and it brings more problems than solutions

:arrow_right: this change is available in my proposition branch: https://github.com/CleverCloud/clever-components/tree/typecheck-i18n--hubert

**Review by hsablonniere — CHANGES_REQUESTED** [review]
Hey @pdesoyres-cc, congrats for this amazing work!!

I had a hard time reading the PR for a few reasons:

* I'm back from holidays :rofl:
* These two changes bring a lot of noise and are mixed with "real changes":
  * Move i18n sources and tests to a dedicated folder and update all components
  * Introduce a stricter `i18n` function (in translations) and update all components

I ended up recreating a git history in a branch to help my understanding of your changes. I also incorporated my suggestions:

* Added some formatting in .storybook (done manually)
* Added some formatting in docs and sandbox (via an update on prettier config)
* Remove LANGUAGE key from translation files (and simplify check i18n and ESLint rules)
* Move export type to cc-beta in another commit
* Use `**/*` patterns in tsconfig

Everything is available in this branch, I'll let you decided if you want to use it or if you want to keep your history as is and apply my suggestions/comments yourself :wink: 

https://github.com/CleverCloud/clever-components/tree/typecheck-i18n--hubert

File: src/lib/i18n/i18n.js [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+ * @typedef {import('./i18n.types.js').TranslateFunction} TranslateFunction
+ * @typedef {import('./i18n.types.js').Translated} Translated
+ */
+
+/**
+ * @param {K} key - The translation key
+ * @param {Object} [data] - The translation data
+ * @returns {Translated} - The translation
+ * @template {string} K
+ */
```
Comment (hsablonniere) [inline src/lib/i18n/i18n.js:13]: issue: I don't think we need the `@template` anymore

:arrow_right: this change is available in my proposition branch: https://github.com/CleverCloud/clever-components/tree/typecheck-i18n--hubert
